tmt/hardware.py

```diff
@@ -0,0 +1,1088 @@
+
+        return ConstraintComponents(
+            property=groups['name'],
+            property_index=int(groups['index']) if groups['index'] is not None else None,
+            child_property=groups['child_name'],
+            operator=groups['operator'],
+            value=groups['value']
+            )
+
+
+def match(text: str, pattern: str) -> bool:
+    """
+    Match a text against a given regular expression.
+
+    :param text: string to examine.
+    :param pattern: regular expression.
+    :returns: ``True`` if pattern matches the string.
+    """
+
+    return re.match(pattern, text) is not None
+
+
+def notmatch(text: str, pattern: str) -> bool:
+    """
+    Match a text against a given regular expression.
+
+    :param text: string to examine.
+    :param pattern: regular expression.
+    :returns: ``True`` if pattern does not matche the string.
+    """
+
+    return re.match(pattern, text) is None
+
+
+def notcontains(haystack: List[str], needle: str) -> bool:
+    """
+    Find out whether an item is in the given list.
+
+    .. note::
+
+       Opposite of :py:func:`operator.contains`.
+
+    :param haystack: container to examine.
+    :param needle: item to look for in ``haystack``.
+    :returns: ``True`` if ``needle`` is in ``haystack``.
+    """
+
+    return needle not in haystack
+
+
+OPERATOR_SIGN_TO_OPERATOR = {
+    '=': Operator.EQ,
+    '==': Operator.EQ,
+    '!=': Operator.NEQ,
+    '>': Operator.GT,
+    '>=': Operator.GTE,
+    '<': Operator.LT,
+    '<=': Operator.LTE,
+    '~': Operator.MATCH,
+    '!~': Operator.NOTMATCH,
+    'contains': Operator.CONTAINS,
+    'not contains': Operator.NOTCONTAINS,
+    # Legacy operators
+    '=~': Operator.MATCH
+    }
+
+
+OPERATOR_TO_HANDLER: Dict[Operator, OperatorHandlerType] = {
+    Operator.EQ: operator.eq,
+    Operator.NEQ: operator.ne,
+    Operator.GT: operator.gt,
+    Operator.GTE: operator.ge,
+    Operator.LT: operator.lt,
+    Operator.LTE: operator.le,
+    Operator.MATCH: match,
+    Operator.NOTMATCH: notmatch,
+    Operator.CONTAINS: operator.contains,
+    Operator.NOTCONTAINS: notcontains
+    }
+
+
+# TODO: provide comment
+ReducerType = Callable[[Iterable[bool]], bool]
+
+
+class ParseError(tmt.utils.MetadataError):
+    """
+    Raised when HW requirement parsing fails.
+    """
+
+    def __init__(self, constraint_name: str, raw_value: str,
+                 message: Optional[str] = None) -> None:
+        """
+        Raise when HW requirement parsing fails.
+
+        :param constraint_name: name of the constraint that caused issues.
+        :param raw_value: original raw value.
+        :param message: optional error message.
+        """
+
+        super().__init__(message or 'failed to parse a constraint')
+
+        self.constraint_name = constraint_name
+        self.raw_value = raw_value
+
+
+#
+# Constraint classes
+#
+
+@dataclasses.dataclass(repr=False)
+class ConstraintBase(tmt.utils.SpecBasedContainer):
```

Comment:
We have `CompoundContraint` defined below, what about calling this `BaseConstraint` to keep consistency?

Reply:
Fixed in e7c0774a.